### PR TITLE
[Delta Protocol] Follow up comments for ClusteringTable feature

### PR DESCRIPTION
**Pull request by dabao521**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Follow up comments for ClusteringTable feature https://github.com/delta-io/delta/pull/2264

We had late comments after the original PR merged, and this PR is addressing those missed comments.

## How was this patch tested?

N/A

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: PROTOCOL.md [hunk 1]
```diff
@@ -1070,14 +1072,35 @@ Enablement:
...
+- Writers must write out [per-file statistics](#per-file-statistics) and per-column statistics for clustering columns in `add` action. 
+  If a new column is included in the clustering columns list, it is required for all table files to have statistics for these added columns.
+- When a clustering implementation clusters files, writers must set the name of the clustering implementation in the `clusteringProvider` field when adding `add` actions for clustered files.
+  - By default, a clustering implementation must only recluster files that have the field `clusteringProvider` set to the name of the same clustering implementation, or to the names of other clustering implementations that are superseded by the current clustering implementation. In addition, a clustering implementation may cluster any files with an unset `clusteringProvider` field (i.e., unclustered files).
+  - Writer is not required to cluster a specific file at any specific moment.
   - A clustering implementation is free to add additional information such as adding a new user-controlled metadata domain to keep track of its metadata.
+- Writers must not define clustered and partitioned table at the same time.
```
Comment (dabao521) [inline PROTOCOL.md:1083]: FYI @ryan-johnson-databricks / @imback82 , this is the new rule to disallow partitioned table . This is to address [comment](https://github.com/delta-io/delta/pull/2264/files#r1393459691) and [comment](https://github.com/delta-io/delta/pull/2264/files#r1393355419)

File: PROTOCOL.md [hunk 2]
```diff
@@ -1070,14 +1072,35 @@ Enablement:
...
+```json
+{
+  "domainMetadata": {
+    "domain": "delta.clustering",
+    "configuration": "{\"clusteringColumns\":[\"col-daadafd7-7c20-4697-98f8-bff70199b1f9\", \"col-5abe0e80-cf57-47ac-9ffc-a861a3d1077e\"]}",
+    "removed": false
+  }
+}
+```
+The example above converts `configuration` field into JSON format, including escaping characters. Here's how it looks in plain JSON for better understanding.
```
Comment (dabao521) [inline PROTOCOL.md:1095]: FYI @ryan-johnson-databricks , this is to address your comment https://github.com/delta-io/delta/pull/2264/files#r1393336240

File: PROTOCOL.md [hunk 3]
```diff
@@ -1057,27 +1059,48 @@ When Row Tracking is enabled (when the table property `delta.enableRowTracking`
 
 The Clustered Table feature facilitates the physical clustering of rows that share similar values on a predefined set of clustering columns.
 This enhances query performance when selective filters are applied to these clustering columns through data skipping.
-Clustering columns must be specified during the initial definition of a clustered table, and they can be modified after the table has been created.
+Clustering columns can be sprecified when creating a table or later, as long as the table doesn't have partition columns.
```
Comment (dabao521) [inline PROTOCOL.md:1062]: FYI @imback82 / @ryan-johnson-databricks , I have updated here and [below](https://github.com/delta-io/delta/pull/2294/files#diff-62d56fa1c25dda8d49ccdbf9ea31df4576f84625706d54613dff0acc8f974d97R1070) to let clustering table feature to be enabled either during table creation or at a later stage

Comment (imback82) [inline PROTOCOL.md:1062]:
```suggestion
Clustering columns can be specified during the initial creation of a table, or they can be added later, provided that the table doesn't have partition columns.
```

**Review by ryan-johnson-databricks — APPROVED** [review]
LGTM

File: PROTOCOL.md (hunk 3 above)
Comment (dabao521) [reply PROTOCOL.md]: fixed

File: PROTOCOL.md [hunk 4]
```diff
@@ -420,6 +421,7 @@ The following is an example `add` action:
     "dataChange": true,
     "baseRowId": 4071,
     "defaultRowCommitVersion": 41,
+    "clusteringProvider": "liquid",
```
Comment (imback82) [inline PROTOCOL.md:424]: this is an invalid scenario where both `partitionValues` and `clusteringProvider` are set?

Comment (dabao521) [reply PROTOCOL.md:439]: Good catch! I added a json example for clusetred table only.

**Review by imback82 — APPROVED** [review]
LGTM

File: PROTOCOL.md (hunk 4 above)
Comment (imback82) [reply PROTOCOL.md:439]: Ah I think we should still have the `"partitionValues":{}`.

Comment (ryan-johnson-databricks) [reply PROTOCOL.md:439]: Confirmed: [AddFile.partitionValues](https://github.com/delta-io/delta/blob/master/spark/src/main/scala/org/apache/spark/sql/delta/actions/actions.scala#L674-L675) is always in the json
```scala
    @JsonInclude(JsonInclude.Include.ALWAYS)
    partitionValues: Map[String, String],
```

Comment (dabao521) [reply PROTOCOL.md:440]: Thanks and I fixed the clustered table json content with an empty partitionValues

**Review by scottsand-db — APPROVED** [review]

**Pull request closed without merging**